PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+                          completion(nil, ErrorUtils.unexpectedBackendResponseError())
+                          return
+                      }
+
+                let paymentDiscount = SKPaymentDiscount(identifier: discountIdentifier,
+                                                        keyIdentifier: keyIdentifier,
+                                                        nonce: nonce,
+                                                        signature: signature,
+                                                        timestamp: timestamp)
+                completion(paymentDiscount, nil)
+            }
+        }
+    }
+
+    @objc(purchaseProduct:payment:presentedOfferingIdentifier:completion:)
+    public func purchase(product: SKProduct,
+                         payment: SKMutablePayment,
+                         presentedOfferingIdentifier maybePresentedOfferingIdentifier: String?,
+                         completion: @escaping PurchaseCompletedBlock) {
+        Logger.debug(String(format: "Make purchase called: %@", #function))
+        guard let productIdentifier = extractProductIdentifier(fromProduct: product, orPayment: payment) else {
+            Logger.error(Strings.purchase.could_not_purchase_product_id_not_found)
+            let errorMessage = "There was a problem purchasing the product: productIdentifier was nil"
+            completion(nil, nil, ErrorUtils.unknownError(message: errorMessage), false)
+            return
+        }
+
+        if !self.finishTransactions {
+            Logger.warn(Strings.purchase.purchasing_with_observer_mode_and_finish_transactions_false_warning)
+        }
+
+        payment.applicationUsername = appUserID
+        preventPurchasePopupCallFromTriggeringCacheRefresh(appUserID: appUserID)
+
+        if let presentedOfferingIdentifier = maybePresentedOfferingIdentifier {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product_from_package,
+                                   productIdentifier,
+                                   presentedOfferingIdentifier))
+            lock.lock()
+            presentedOfferingIDsByProductID[productIdentifier] = presentedOfferingIdentifier
+            lock.unlock()
+
+        } else {
+            Logger.purchase(String(format: Strings.purchase.purchasing_product, productIdentifier))
+        }
+
+        productsManager.cacheProduct(product)
+
+        lock.lock()
+        defer {
+            lock.unlock()
+        }
+
+        guard purchaseCompleteCallbacksByProductID[productIdentifier] == nil else {
+            completion(nil, nil, ErrorUtils.operationAlreadyInProgressError(), false)
+            return
+        }
+        purchaseCompleteCallbacksByProductID[productIdentifier] = completion
+        storeKitWrapper.add(payment)
+    }
+
+}
+
+extension PurchasesOrchestrator: StoreKitWrapperDelegate {
+
+    public func storeKitWrapper(_ storeKitWrapper: StoreKitWrapper,
```

Comment:
probably not part of this PR, but wondering if this function name could be clearer. it sounds more like some kind of initializer rather than something that is performing an action...maybe `handleTransaction`?

Reply:
maybe i'm missing some convention here

Reply:
There is some convention here that's totally non-obvious. If you end up passing an object that is performing an action in a delegate, it's common to start the function name with the name of the object. 
In `StoreKitWrapperDelegate`, we have 
`storeKitWrapper(_ storeKitWrapper: StoreKitWrapper, updatedTransaction transaction: SKPaymentTransaction)`
Which, in english reads "StoreKitWrapper `x` just updated this transaction `y`"
You can see this pattern in action with tableviews:
https://developer.apple.com/documentation/uikit/uitableviewdelegate/1614877-tableview

TableViewDelegate has:
`func tableView(_ tableView: UITableView,  didSelectRowAt indexPath: IndexPath)`
"TableView `x` just had row selected at indexPath `y`"

Again, I only see this when the first parameter of the delegate method is actually the object that is performing the action.